File: project/android/apps/MnnLlmApp/app/src/main/java/com/alibaba/mnnllm/android/ChatSession.java
// Created by ruoyi.sjd on 2024/12/25.
// Copyright (c) 2024 Alibaba Group Holding Limited All rights reserved.

package com.alibaba.mnnllm.android;

import android.util.Log;

import com.alibaba.mnnllm.android.chat.ChatDataItem;
import com.alibaba.mnnllm.android.utils.ModelUtils;

import java.io.Serializable;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

public class ChatSession implements Serializable {

    private final String configPath;
    private long nativePtr;

    public static final String TAG = "ChatSession";

    private String sessionId;

    private volatile boolean mGenerating = false;
    private volatile boolean mReleaseRequeted = false;

    private List<ChatDataItem> savedHistory;
    private boolean isDiffusion;

    private boolean useTmpPath;
    private boolean keepHistory;

    public ChatSession(String sessionId, String configPath, boolean useTmpPath, List<ChatDataItem> history) {
        this(sessionId, configPath, useTmpPath, history, false);
    }

    public ChatSession(String sessionId, String configPath, boolean useTmpPath, List<ChatDataItem> history, boolean isDiffusion) {
        this.sessionId = sessionId;
        this.configPath = configPath;
        this.savedHistory = history;
        this.isDiffusion = isDiffusion;
        this.useTmpPath = useTmpPath;
    }

    public void load() {
        Log.d(TAG, "MNN_DEBUG load begin");
        List<String> historyStringList = null;
        if (this.savedHistory != null && !this.savedHistory.isEmpty()) {
            historyStringList = this.savedHistory.stream().map(ChatDataItem::getText).collect(Collectors.toList());
        }
        nativePtr = initNative(configPath, useTmpPath, historyStringList, isDiffusion);
    }

    public List<ChatDataItem> getSavedHistory() {
        return savedHistory;
    }

    public String generateNewSession() {
        this.sessionId = String.valueOf(System.currentTimeMillis());
        return this.sessionId;
    }

    public String getSessionId() {
        return sessionId;
    }

    public HashMap<String, Object> generate(String input, GenerateProgressListener progressListener) {
        synchronized (this) {
            Log.d(TAG, "MNN_DEBUG submit" + input);
            mGenerating = true;
            HashMap<String, Object> result = submitNative(nativePtr, input, keepHistory, progressListener);
            mGenerating = false;
            if (mReleaseRequeted) {
                releaseInner();
            }
            return result;
        }
    }

    public HashMap<String, Object> generateDiffusion(String input, String output, GenerateProgressListener progressListener) {
        synchronized (this) {
            Log.d(TAG, "MNN_DEBUG submit" + input);
            mGenerating = true;
            HashMap<String, Object> result = submitDiffusionNative(nativePtr, input, output, progressListener);
            mGenerating = false;
            if (mReleaseRequeted) {
                releaseInner();
            }
            return result;
        }
    }

    public void reset() {
        synchronized (this) {
            resetNative(nativePtr);
        }
    }

    public void release() {
        synchronized (this) {
            Log.d(TAG, "MNN_DEBUG release" + nativePtr);
            if (!mGenerating) {
                releaseInner();
            } else {
                mReleaseRequeted = true;
            }
        }
    }

    private void releaseInner() {
        if (nativePtr > 0) {
            releaseNative(nativePtr, isDiffusion);
            nativePtr = 0;
            ChatService.provide().removeSession(sessionId);
        }
    }

    @Override
    protected void finalize() throws Throwable {
        super.finalize();
        release();
    }

    public native long initNative(String configPath, boolean useTmpPath, List<String> history, boolean isDiffusion);
    private native HashMap<String, Object> submitNative(long instanceId, String input, boolean keepHistory, GenerateProgressListener listener);

    private native HashMap<String, Object> submitDiffusionNative(long instanceId, String input, String outputPath, GenerateProgressListener progressListener);
    private native void resetNative(long instanceId);

    private native void releaseNative(long instanceId, boolean isDiffusion);

    static {
        System.loadLibrary("llm");
        System.loadLibrary("MNN_CL");
    }

    public void setKeepHistory(boolean keepHistory) {
        this.keepHistory = keepHistory;
    }

    public interface GenerateProgressListener {
        boolean onProgress(String progress);
    }
}

File: apps/Android/MnnTaoAvatar/app/src/main/java/com/alibaba/mnnllm/android/ChatSession.kt
// Created by ruoyi.sjd on 2024/12/25.
// Copyright (c) 2024 Alibaba Group Holding Limited All rights reserved.
package com.alibaba.mnnllm.android

import android.util.Log
import com.alibaba.mls.api.ApplicationProvider
import com.alibaba.mnnllm.android.ChatService.Companion.provide
import com.alibaba.mnnllm.android.chat.ChatDataItem
import com.alibaba.mnnllm.android.utils.FileUtils.getMmapDir
import com.alibaba.mnnllm.android.utils.ModelPreferences.useMmap
import com.alibaba.mnnllm.android.utils.ModelUtils.isR1Model
import com.taobao.meta.avatar.settings.MainSettings.getLlmPrompt
import java.io.File
import java.io.Serializable
import java.util.stream.Collectors
import kotlin.concurrent.Volatile

class ChatSession @JvmOverloads constructor(
    private val modelId: String,
    var sessionId: String,
    private val configPath: String,
    private val useTmpPath: Boolean,
    val savedHistory: List<ChatDataItem>?,
    private val isDiffusion: Boolean = false
) :
    Serializable {
    private var nativePtr: Long = 0

    @Volatile
    private var modelLoading = false

    @Volatile
    private var mGenerating = false

    @Volatile
    private var mReleaseRequeted = false

    private var keepHistory = false

    fun load() {
        Log.d(TAG, "MNN_DEBUG load begin")
        modelLoading = true
        var historyStringList: List<String>? = null
        if (this.savedHistory != null && !savedHistory.isEmpty()) {
            historyStringList =
                savedHistory.stream().map { obj: ChatDataItem -> obj.text }
                    .collect(Collectors.toList())
        }
        var rootCacheDir = ""
        if (useMmap(ApplicationProvider.get(), modelId)) {
            rootCacheDir = getMmapDir(modelId, configPath.contains("modelscope"))
            File(rootCacheDir).mkdirs()
        }
        rootCacheDir = ApplicationProvider.get().cacheDir.toString() + "/llm"
        File(rootCacheDir).mkdirs()
        val extraParams = HashMap<String, Any>()
        extraParams["system_prompt"] = getLlmPrompt(ApplicationProvider.get())
        nativePtr = initNative(
            "", modelId, configPath, useTmpPath,
            historyStringList, isDiffusion, isR1Model(modelId), extraParams
        )
        modelLoading = false
        if (mReleaseRequeted) {
            release()
        }
    }

    val debugInfo: String
        get() = getDebugInfoNative(nativePtr) + "\n"

    fun generateNewSession(): String {
        this.sessionId = System.currentTimeMillis().toString()
        return this.sessionId
    }

    fun generate(input: String, progressListener: GenerateProgressListener): HashMap<String, Any> {
        synchronized(this) {
            Log.d(TAG, "MNN_DEBUG submit$input")
            mGenerating = true
            val result = submitNative(nativePtr, input, keepHistory, progressListener)
            mGenerating = false
            if (mReleaseRequeted) {
                release()
            }
            return result
        }
    }

    fun generateDiffusion(
        input: String,
        output: String?,
        progressListener: GenerateProgressListener?
    ): HashMap<String, Any> {
        synchronized(this) {
            Log.d(TAG, "MNN_DEBUG submit$input")
            mGenerating = true
            val result = HashMap<String, Any>()
            return result
        }
    }

    fun reset() {
        synchronized(this) {
            resetNative(nativePtr)
        }
    }

    fun release() {
        synchronized(this) {
            Log.d(
                TAG,
                "MNN_DEBUG release nativePtr: $nativePtr mGenerating: $mGenerating"
            )
            if (!mGenerating && !modelLoading) {
                releaseInner()
            } else {
                mReleaseRequeted = true
                while (mGenerating || modelLoading) {
                    try {
                        (this as Object).wait()
                    } catch (e: InterruptedException) {
                        Thread.currentThread().interrupt()
                        Log.e(TAG, "Thread interrupted while waiting for release", e)
                    }
                }
                releaseInner()
            }
        }
    }

    private fun releaseInner() {
        if (nativePtr != 0L) {
            releaseNative(nativePtr, isDiffusion)
            nativePtr = 0
            provide().removeSession(sessionId)
            (this as Object).notifyAll()
        }
    }

    external fun initNative(
        rootCacheDir: String?,
        modelId: String?,
        configPath: String?,
        useTmpPath: Boolean,
        history: List<String>?,
        isDiffusion: Boolean,
        isR1: Boolean,
        extraParams: HashMap<String, Any>?
    ): Long

    private external fun submitNative(
        instanceId: Long,
        input: String,
        keepHistory: Boolean,
        listener: GenerateProgressListener
    ): HashMap<String, Any>

    private external fun resetNative(instanceId: Long)

    private external fun getDebugInfoNative(instanceId: Long): String

    private external fun releaseNative(instanceId: Long, isDiffusion: Boolean)

    fun setKeepHistory(keepHistory: Boolean) {
        this.keepHistory = keepHistory
    }

    fun updatePrompt(llmPrompt: String) {
        val extraParams = HashMap<String, Any>()
        extraParams["system_prompt"] = llmPrompt
        updateConfigNative(nativePtr, extraParams)
    }

    external fun updateConfigNative(nativePtr: Long, extraParams: HashMap<String, Any>?)
    interface GenerateProgressListener {
        fun onProgress(progress: String?): Boolean
    }

    companion object {

        const val TAG: String = "ChatSession"
    }
}
